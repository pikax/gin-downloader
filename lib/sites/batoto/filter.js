"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Created by rodriguesc on 30/03/2017.
 */
var declarations_1 = require("../../declarations");
var config_1 = require("./config");
var Supported = [];
Supported[declarations_1.Genre.Action] = declarations_1.Genre.Action;
Supported[declarations_1.Genre.Adventure] = declarations_1.Genre.Adventure;
Supported[declarations_1.Genre.Comedy] = declarations_1.Genre.Comedy;
Supported[declarations_1.Genre.Mystery] = declarations_1.Genre.Mystery;
Supported[declarations_1.Genre.Psychological] = declarations_1.Genre.Psychological;
Supported[declarations_1.Genre.Romance] = declarations_1.Genre.Romance;
Supported[declarations_1.Genre.SchoolLife] = declarations_1.Genre.SchoolLife;
Supported[declarations_1.Genre.SciFi] = declarations_1.Genre.SciFi;
Supported[declarations_1.Genre.Doujinshi] = declarations_1.Genre.Doujinshi;
Supported[declarations_1.Genre.Drama] = declarations_1.Genre.Drama;
Supported[declarations_1.Genre.Ecchi] = declarations_1.Genre.Ecchi;
Supported[declarations_1.Genre.Fantasy] = declarations_1.Genre.Fantasy;
Supported[declarations_1.Genre.GenderBender] = declarations_1.Genre.GenderBender;
Supported[declarations_1.Genre.ShoujoAi] = declarations_1.Genre.ShoujoAi;
Supported[declarations_1.Genre.Harem] = declarations_1.Genre.Harem;
Supported[declarations_1.Genre.ShounenAi] = declarations_1.Genre.ShounenAi;
Supported[declarations_1.Genre.Historical] = declarations_1.Genre.Historical;
Supported[declarations_1.Genre.SliceOfLife] = declarations_1.Genre.SliceOfLife;
Supported[declarations_1.Genre.Horror] = declarations_1.Genre.Horror;
Supported[declarations_1.Genre.Smut] = declarations_1.Genre.Smut;
Supported[declarations_1.Genre.Sports] = declarations_1.Genre.Sports;
Supported[declarations_1.Genre.Supernatural] = declarations_1.Genre.Supernatural;
Supported[declarations_1.Genre.MartialArts] = declarations_1.Genre.MartialArts;
Supported[declarations_1.Genre.Tragedy] = declarations_1.Genre.Tragedy;
Supported[declarations_1.Genre.Yaoi] = declarations_1.Genre.Yaoi;
Supported[declarations_1.Genre.Mecha] = declarations_1.Genre.Mecha;
Supported[declarations_1.Genre.Yuri] = declarations_1.Genre.Yuri;
Supported[declarations_1.Genre.Seinen] = declarations_1.Genre.Seinen;
Supported[declarations_1.Genre.Shounen] = declarations_1.Genre.Shounen;
Supported[declarations_1.Genre.Josei] = declarations_1.Genre.Josei;
Supported[declarations_1.Genre.Shoujo] = declarations_1.Genre.Shoujo;
Supported[declarations_1.Genre.Webtoon] = declarations_1.Genre.Webtoon;
Supported[declarations_1.Genre.Music] = declarations_1.Genre.Music;
Supported[declarations_1.Genre.Oneshot] = declarations_1.Genre.Oneshot;
Supported[declarations_1.Genre.AwardWinning] = declarations_1.Genre.AwardWinning;
Supported[declarations_1.Genre.FourKoma] = declarations_1.Genre.FourKoma;
Supported[declarations_1.Genre.Cooking] = declarations_1.Genre.Cooking;
Supported[declarations_1.Genre.Medical] = declarations_1.Genre.Medical;
Supported[declarations_1.Genre.NoChapters] = declarations_1.Genre.NoChapters;
var ordered = [
    declarations_1.Genre.Action,
    declarations_1.Genre.Adventure,
    declarations_1.Genre.Comedy,
    declarations_1.Genre.Mystery,
    declarations_1.Genre.Psychological,
    declarations_1.Genre.Romance,
    declarations_1.Genre.SchoolLife,
    declarations_1.Genre.SciFi,
    declarations_1.Genre.Doujinshi,
    declarations_1.Genre.Drama,
    declarations_1.Genre.Ecchi,
    declarations_1.Genre.Fantasy,
    declarations_1.Genre.GenderBender,
    declarations_1.Genre.ShoujoAi,
    declarations_1.Genre.Harem,
    declarations_1.Genre.ShounenAi,
    declarations_1.Genre.Historical,
    declarations_1.Genre.SliceOfLife,
    declarations_1.Genre.Horror,
    declarations_1.Genre.Smut,
    declarations_1.Genre.Sports,
    declarations_1.Genre.Supernatural,
    declarations_1.Genre.MartialArts,
    declarations_1.Genre.Tragedy,
    declarations_1.Genre.Yaoi,
    declarations_1.Genre.Mecha,
    declarations_1.Genre.Yuri,
    declarations_1.Genre.Seinen,
    declarations_1.Genre.Shounen,
    declarations_1.Genre.Josei,
    declarations_1.Genre.Shoujo,
    declarations_1.Genre.Webtoon,
    declarations_1.Genre.Music,
    declarations_1.Genre.Oneshot,
    declarations_1.Genre.AwardWinning,
    declarations_1.Genre.FourKoma,
    declarations_1.Genre.Cooking,
    declarations_1.Genre.Medical,
    declarations_1.Genre.NoChapters,
];
var dic = {};
dic[declarations_1.Genre.Action] = "1";
dic[declarations_1.Genre.Adventure] = "2";
dic[declarations_1.Genre.Comedy] = "3";
dic[declarations_1.Genre.Mystery] = "4";
dic[declarations_1.Genre.Psychological] = "5";
dic[declarations_1.Genre.Romance] = "6";
dic[declarations_1.Genre.SchoolLife] = "7";
dic[declarations_1.Genre.SciFi] = "8";
dic[declarations_1.Genre.Doujinshi] = "9";
dic[declarations_1.Genre.Drama] = "10";
dic[declarations_1.Genre.Ecchi] = "12";
dic[declarations_1.Genre.Fantasy] = "13";
dic[declarations_1.Genre.GenderBender] = "15";
dic[declarations_1.Genre.ShoujoAi] = "16";
dic[declarations_1.Genre.Harem] = "17";
dic[declarations_1.Genre.ShounenAi] = "19";
dic[declarations_1.Genre.Historical] = "20";
dic[declarations_1.Genre.SliceOfLife] = "21";
dic[declarations_1.Genre.Horror] = "22";
dic[declarations_1.Genre.Smut] = "23";
dic[declarations_1.Genre.Sports] = "25";
dic[declarations_1.Genre.Supernatural] = "26";
dic[declarations_1.Genre.MartialArts] = "27";
dic[declarations_1.Genre.Tragedy] = "28";
dic[declarations_1.Genre.Yaoi] = "29";
dic[declarations_1.Genre.Mecha] = "30";
dic[declarations_1.Genre.Yuri] = "31";
dic[declarations_1.Genre.Seinen] = "32";
dic[declarations_1.Genre.Shounen] = "33";
dic[declarations_1.Genre.Josei] = "34";
dic[declarations_1.Genre.Shoujo] = "35";
dic[declarations_1.Genre.Webtoon] = "36";
dic[declarations_1.Genre.Music] = "37";
dic[declarations_1.Genre.Oneshot] = "38";
dic[declarations_1.Genre.AwardWinning] = "39";
dic[declarations_1.Genre.FourKoma] = "40";
dic[declarations_1.Genre.Cooking] = "41";
dic[declarations_1.Genre.Medical] = "42";
dic[declarations_1.Genre.NoChapters] = "44";
exports.processFilter = function (filter) {
    filter = filter || {};
    var genres = filter.genres, outGenres = filter.outGenres, search = filter.search;
    var fauthor = null;
    var fstatus = null;
    var ftype = null;
    var fname = filter.name;
    var nameCondition = null;
    var authorCondition = null;
    var genreCondition = null;
    if (search) {
        var nameFilter = search.name;
        if (nameFilter) {
            fname = nameFilter.name;
            nameCondition = resolveCondition(nameFilter.condition); // not support equals just Starts|Ends|Contains
        }
        var authorFilter = search.author || search.artist;
        if (authorFilter) {
            fauthor = authorFilter.name;
            authorCondition = resolveCondition(authorFilter.condition);
        }
        var statusFilter = search.status;
        if (statusFilter) {
            fstatus = resolveStatus(statusFilter);
        }
        var genreFilter_1 = search.genre;
        if (genreFilter_1) {
            genres = genreFilter_1.inGenres;
            outGenres = genreFilter_1.outGenres;
            genreCondition = resolveGenreCondition(genreFilter_1.condition);
        }
        var type_1 = search.type;
        if (type_1) {
            ftype = resolveType(type_1);
        }
    }
    var mangaName = "name=" + (fname || "");
    var nameCond = nameCondition && "name_cond=" + nameCondition;
    var authorArtist = "artist_name=" + (fauthor || "");
    var authorCond = authorCondition && "artist_name_cond=" + authorCondition;
    var genreFilter = "genres=" + ordered.map(function (x) { return inOutGenre(x, genres, outGenres); }).filter(function (x) { return x !== ""; }).join(";");
    var genreCon = (genreCondition && "genre_cond=" + genreCondition) || "genre_cond=and"; // todo change me
    var status = "status=" + (fstatus || "");
    var type = "type=" + ftype;
    var page = "p=" + (filter.page || 1);
    // TODO implement me
    var rating_high = 0; // 0~5
    var rating_low = 0; // 0~5
    var mature = false; // n == false
    return { src: config_1.config.mangas_url + "?" + [mangaName, nameCond, authorArtist, authorCond, genreFilter, genreCon, status, type, page].join("&")
    };
};
function resolveType(type) {
    switch (type) {
        case declarations_1.FilterMangaType.Manga:
            return "jp";
        case declarations_1.FilterMangaType.Manhwa:
            return "kr";
        case declarations_1.FilterMangaType.Manhua:
            return "cn";
        case declarations_1.FilterMangaType.Artbook:
            return "ar";
        case declarations_1.FilterMangaType.Other:
            return "ot";
        default:
            throw new Error("Unknown type");
    }
}
function resolveCondition(condition) {
    switch (condition) {
        case declarations_1.FilterCondition.Contains:
            return "c";
        case declarations_1.FilterCondition.EndsWith:
            return "e";
        case declarations_1.FilterCondition.StartsWith:
            return "s";
        case declarations_1.FilterCondition.Equal:
            return "is";
        default:
            return "c";
    }
}
function resolveGenreCondition(condition) {
    switch (condition) {
        case declarations_1.GenreCondition.And:
            return "and";
        case declarations_1.GenreCondition.Or:
            return "or";
        default:
            return "and";
    }
}
function resolveStatus(status) {
    switch (status) {
        case declarations_1.FilterStatus.Ongoing:
            return "i";
        case declarations_1.FilterStatus.Complete:
            return "c";
        default:
            return null;
    }
}
function inOutGenre(genre, inGenre, outGenre) {
    if (inGenre && inGenre.indexOf(genre) > -1) {
        return "i" + dic[genre];
    }
    if (outGenre && outGenre.indexOf(genre) > -1) {
        return "e" + dic[genre];
    }
    return "";
}
//# sourceMappingURL=filter.js.map