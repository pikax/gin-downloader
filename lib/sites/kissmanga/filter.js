"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Created by rodriguesc on 30/03/2017.
 */
var declarations_1 = require("../../declarations");
var config_1 = require("./config");
var url_1 = require("url");
var Supported = [];
Supported[declarations_1.Genre.FourKoma] = declarations_1.Genre.FourKoma;
Supported[declarations_1.Genre.Action] = declarations_1.Genre.Action;
Supported[declarations_1.Genre.Adult] = declarations_1.Genre.Adult;
Supported[declarations_1.Genre.Adventure] = declarations_1.Genre.Adventure;
Supported[declarations_1.Genre.Comedy] = declarations_1.Genre.Comedy;
Supported[declarations_1.Genre.Comic] = declarations_1.Genre.Comic;
Supported[declarations_1.Genre.Cooking] = declarations_1.Genre.Cooking;
Supported[declarations_1.Genre.Doujinshi] = declarations_1.Genre.Doujinshi;
Supported[declarations_1.Genre.Drama] = declarations_1.Genre.Drama;
Supported[declarations_1.Genre.Ecchi] = declarations_1.Genre.Ecchi;
Supported[declarations_1.Genre.Fantasy] = declarations_1.Genre.Fantasy;
Supported[declarations_1.Genre.GenderBender] = declarations_1.Genre.GenderBender;
Supported[declarations_1.Genre.Harem] = declarations_1.Genre.Harem;
Supported[declarations_1.Genre.Historical] = declarations_1.Genre.Historical;
Supported[declarations_1.Genre.Horror] = declarations_1.Genre.Horror;
Supported[declarations_1.Genre.Josei] = declarations_1.Genre.Josei;
Supported[declarations_1.Genre.Lolicon] = declarations_1.Genre.Lolicon;
Supported[declarations_1.Genre.Manga] = declarations_1.Genre.Manga;
Supported[declarations_1.Genre.Manhua] = declarations_1.Genre.Manhua;
Supported[declarations_1.Genre.Manhwa] = declarations_1.Genre.Manhwa;
Supported[declarations_1.Genre.MartialArts] = declarations_1.Genre.MartialArts;
Supported[declarations_1.Genre.Mature] = declarations_1.Genre.Mature;
Supported[declarations_1.Genre.Mecha] = declarations_1.Genre.Mecha;
Supported[declarations_1.Genre.Medical] = declarations_1.Genre.Medical;
Supported[declarations_1.Genre.Music] = declarations_1.Genre.Music;
Supported[declarations_1.Genre.Mystery] = declarations_1.Genre.Mystery;
Supported[declarations_1.Genre.Oneshot] = declarations_1.Genre.Oneshot;
Supported[declarations_1.Genre.Psychological] = declarations_1.Genre.Psychological;
Supported[declarations_1.Genre.Romance] = declarations_1.Genre.Romance;
Supported[declarations_1.Genre.SchoolLife] = declarations_1.Genre.SchoolLife;
Supported[declarations_1.Genre.SciFi] = declarations_1.Genre.SciFi;
Supported[declarations_1.Genre.Seinen] = declarations_1.Genre.Seinen;
Supported[declarations_1.Genre.Shotacon] = declarations_1.Genre.Shotacon;
Supported[declarations_1.Genre.Shoujo] = declarations_1.Genre.Shoujo;
Supported[declarations_1.Genre.ShoujoAi] = declarations_1.Genre.ShoujoAi;
Supported[declarations_1.Genre.Shounen] = declarations_1.Genre.Shounen;
Supported[declarations_1.Genre.ShounenAi] = declarations_1.Genre.ShounenAi;
Supported[declarations_1.Genre.SliceOfLife] = declarations_1.Genre.SliceOfLife;
Supported[declarations_1.Genre.Smut] = declarations_1.Genre.Smut;
Supported[declarations_1.Genre.Sports] = declarations_1.Genre.Sports;
Supported[declarations_1.Genre.Supernatural] = declarations_1.Genre.Supernatural;
Supported[declarations_1.Genre.Tragedy] = declarations_1.Genre.Tragedy;
Supported[declarations_1.Genre.Webtoon] = declarations_1.Genre.Webtoon;
Supported[declarations_1.Genre.Yaoi] = declarations_1.Genre.Yaoi;
Supported[declarations_1.Genre.Yuri] = declarations_1.Genre.Yuri;
var ordered = [
    declarations_1.Genre.FourKoma,
    declarations_1.Genre.Action,
    declarations_1.Genre.Adult,
    declarations_1.Genre.Adventure,
    declarations_1.Genre.Comedy,
    declarations_1.Genre.Comic,
    declarations_1.Genre.Cooking,
    declarations_1.Genre.Doujinshi,
    declarations_1.Genre.Drama,
    declarations_1.Genre.Ecchi,
    declarations_1.Genre.Fantasy,
    declarations_1.Genre.GenderBender,
    declarations_1.Genre.Harem,
    declarations_1.Genre.Historical,
    declarations_1.Genre.Horror,
    declarations_1.Genre.Josei,
    declarations_1.Genre.Lolicon,
    declarations_1.Genre.Manga,
    declarations_1.Genre.Manhua,
    declarations_1.Genre.Manhwa,
    declarations_1.Genre.MartialArts,
    declarations_1.Genre.Mature,
    declarations_1.Genre.Mecha,
    declarations_1.Genre.Medical,
    declarations_1.Genre.Music,
    declarations_1.Genre.Mystery,
    declarations_1.Genre.Oneshot,
    declarations_1.Genre.Psychological,
    declarations_1.Genre.Romance,
    declarations_1.Genre.SchoolLife,
    declarations_1.Genre.SciFi,
    declarations_1.Genre.Seinen,
    declarations_1.Genre.Shotacon,
    declarations_1.Genre.Shoujo,
    declarations_1.Genre.Shoujo,
    declarations_1.Genre.Shounen,
    declarations_1.Genre.Shounen,
    declarations_1.Genre.SliceOfLife,
    declarations_1.Genre.Smut,
    declarations_1.Genre.Sports,
    declarations_1.Genre.Supernatural,
    declarations_1.Genre.Tragedy,
    declarations_1.Genre.Webtoon,
    declarations_1.Genre.Yaoi,
    declarations_1.Genre.Yuri,
];
exports.processFilter = function (filter) {
    filter = filter || {};
    var genres = filter.genres, outGenres = filter.outGenres;
    var mangaName = "mangaName=" + (filter.name || "");
    var authorArtist = "authorArtist=";
    var genreFilter = ordered.map(function (x) { return inOutGenre(x, genres, outGenres); }).map(function (x) { return "genres=" + x; }).join("&");
    var status = "status=";
    return { src: url_1.resolve(config_1.config.site, "/AdvanceSearch"),
        params: [mangaName, authorArtist, genreFilter, status].join("&")
    };
};
function inOutGenre(genre, inGenre, outGenre) {
    if (inGenre && inGenre.indexOf(genre) > -1) {
        return 1;
    }
    if (outGenre && outGenre.indexOf(genre) > -1) {
        return 2;
    }
    return 0;
}
//# sourceMappingURL=filter.js.map