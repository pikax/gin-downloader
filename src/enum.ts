export enum Genre {
    Action = "Action",
    Adult = "Adult",
    Adventure = "Adventure",
    AwardWinning = "Award Winning",
    Comedy = "Comedy",
    Comic = "Comic",
    Cooking = "Cooking",
    Demons = "Demons",
    Doujinshi = "Doujinshi",
    Drama = "Drama",
    Ecchi = "Ecchi",
    Fantasy = "Fantasy",
    FourKoma = "4-Koma",
    GenderBender = "Gender Bender",
    Harem = "Harem",
    Historical = "Historical",
    Horror = "Horror",
    Josei = "Josei",
    Lolicon = "Lolicon",
    Magic = "Magic",
    Manga = "Manga",
    Manhua = "Manhua",
    Manhwa = "Manhwa",
    MartialArts = "Martial Arts",
    Mature = "Mature",
    Mecha = "Mecha",
    Medical = "Medical",
    Military = "Military",
    Music = "Music",
    Mystery = "Mystery",
    Oneshot = "Oneshot",
    Psychological = "Psychological",
    Romance = "Romance",
    SchoolLife = "School Life",
    SciFi = "Sci-fi",
    Seinen = "Seinen",
    Shotacon = "Shotacon",
    Shoujo = "Shoujo",
    ShoujoAi = "Shoujo Ai",
    Shounen = "Shounen",
    ShounenAi = "Shounen Ai",
    SliceOfLife = "Slice of Life",
    Smut = "Smut",
    Sports = "Sports",
    Supernatural = "Supernatural",
    SuperPower = "SuperPower",
    Tragedy = "Tragedy",
    Vampire = "Vampire",
    Webtoon = "Webtoon",
    Yaoi = "Yaoi",
    Yuri = "Yuri",

    NoChapters = "[no chapters]"
}

export enum Type {
    Manga = "Manga",
    Manhwa = "Manhwa",
    Manhua = "Manhua",
    Comic = "Comic",
    Artbook = "Artbook", // An artbook is a title that contains purely art and has no story
    Other = "Other", // bato.to
}


export enum GenreCondition {
    And = 0,
    Or = 1,
}

export enum FilterStatus {
    Ongoing = "Ongoing",
    Complete = "Complete",
    Cancelled = "Cancelled"
}


export enum FilterCondition {
    Equal,
    Contains,
    NotContains,
    StartsWith,
    EndsWith,
    Less,
    Greater,
    LessThan,
    GreaterThan,
}