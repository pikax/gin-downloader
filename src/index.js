/**
 * Created by david on 19/03/2017.
 */

import fs from 'fs';

import builder from './sites';


let sites = {};

const sitesDir = __dirname + '/sites/';
fs.readdirSync(sitesDir).forEach((file)=> {
  let fp = sitesDir + file;

  if( fs.statSync(fp).isFile())
    return; //nothing to do here

  let site = require(fp).default;
  let parser = require(fp + '/parser').default;
  let config = require(fp + '/config').default;


  let def = new builder(site,config,parser);

  let s = site;

  s.mangas = s.mangas || def.mangas.bind(def);

  site[config.name] = s;
});


modules.export = site;


