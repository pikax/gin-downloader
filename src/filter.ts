import {FilterCondition, FilterStatus, Genre, GenreCondition, Type} from "./enum";
import {IManga, IMangaObject, Source} from "./interface";

export type GenreCollection = Array<Genre>;

export interface NameFilterCondition {
    name: string;
    condition?: FilterCondition;
}

export interface ValueFilter {
    value: number;
    condition?: FilterCondition;
}

export interface GenreFilter {
    inGenres?: GenreCollection;
    outGenres?: GenreCollection;
    condition?: GenreCondition;
}

export interface RatingFilter { // from
    from?: number;
    to?: number;
}

export type NameFilter = NameFilterCondition;

export type AuthorFilter = NameFilter;
export type ArtistFilter = NameFilter;
export type ReleaseFilter = ValueFilter;

export interface Search {
    name?: NameFilter | string;
    author?: AuthorFilter | string;
    artist?: ArtistFilter | string;
    status?: FilterStatus | string;
    released?: ReleaseFilter | number;
    genre?: GenreFilter | GenreCollection;
    rating?: RatingFilter| number;
    mature?: boolean;

    type?: Type;
}

export interface MangaFilter {
    name?: string;

    search?: Search;
    sort?: {};

    page?: number;
}

export type MangaSource = IManga & Source;

export interface FilteredResults {
    results: IMangaObject[];
    page: number;
    total: number;
}

export namespace filter {


    export interface Search {
        name?: NameFilter;
        author?: AuthorFilter;
        artist?: ArtistFilter;
        status?: FilterStatus;
        released?: ReleaseFilter;
        genre?: GenreFilter;
        rating?: RatingFilter;
        mature?: boolean;

        type?: Type;
    }

    export interface FilterSupport {
        search?: Search;
        sort?: {};

        page?: number;
    }
}
