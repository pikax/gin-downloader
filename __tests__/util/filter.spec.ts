import {sanitizeFilter} from "../../src/util/filter";
import {MangaFilter} from "../../src/filter";
import {FilterCondition, FilterStatus, Genre, GenreCondition, Type} from "../../src/enum";


describe("filter util", () => {


    it("should sanitize filter: string", () => {
        const filter = "gintama";

        const filterSupport = sanitizeFilter(filter);

        expect(filterSupport).toMatchSnapshot();
    });

    it("should sanitize filter: user search ", () => {

        const dirtyFilter: MangaFilter = {
            search: {
                genre: [Genre.Shounen, Genre.Action],
                artist: "Sorachi Hideaki",
                author: "Sorachi Hideaki",
                mature: true,
                name: "gintama",
                rating: 1,
                released: 2000,
                status: FilterStatus.Ongoing,
                type: Type.Manga
            }
        };


        const filter = sanitizeFilter(dirtyFilter);

        expect(filter).toMatchSnapshot();
    });

    it("should sanitize filter: full", () => {
        const dirtyFilter: MangaFilter = {
            search: {
                genre: {
                    inGenres: [Genre.Shounen, Genre.Action],
                    outGenres: [Genre.Comic],
                    condition: GenreCondition.Or
                },
                artist: {name: "Sorachi Hideaki", condition: FilterCondition.Contains},
                author: {name: "Sorachi Hideaki", condition: FilterCondition.Contains},
                mature: true,
                name: {name: "gintama", condition: FilterCondition.Contains},
                rating: {from: 1, to: 4},
                released: {value: 2000, condition: FilterCondition.Equal},
                status: FilterStatus.Ongoing,
                type: Type.Manga
            }
        };

        const filter = sanitizeFilter(dirtyFilter);

        // they should be the same
        expect(filter).toEqual(dirtyFilter);
    });

    describe("invalid inputs", () => {

        describe("Genre", () => {
            it("by Genre", () => {
                const fn = () => sanitizeFilter(null);

                expect(fn).not.toThrow();
                expect(fn()).toMatchSnapshot();
            });

            it("by inGenre", () => {
                const condition: any = {
                    search: {
                        genre: {
                            inGenres: false
                        }
                    }
                };
                const fn = () => sanitizeFilter(condition);

                expect(fn).not.toThrow();
                expect(fn()).toMatchSnapshot();
            });

            it("by outGenre", () => {
                const condition: any = {
                    search: {
                        genre: {
                            outGenres: false
                        }
                    }
                };
                const fn = () => sanitizeFilter(condition);

                expect(fn).not.toThrow();
                expect(fn()).toMatchSnapshot();
            });

        });

        it("by Artist", () => {
            const condition: any = {
                search: {
                    artist: true
                }
            };
            const fn = () => sanitizeFilter(condition);

            expect(fn).not.toThrow();
            expect(fn()).toMatchSnapshot();
        });

        it("by Author", () => {
            const condition: any = {
                search: {
                    author: true
                }
            };
            const fn = () => sanitizeFilter(condition);

            expect(fn).not.toThrow();
            expect(fn()).toMatchSnapshot();
        });

        it("by Mature", () => {
            const condition: any = {
                search: {
                    mature: {value: true}
                }
            };
            const fn = () => sanitizeFilter(condition);

            expect(fn).not.toThrow();
            expect(fn()).toMatchSnapshot();
        });

        describe("Name", () => {
            it("by Name", () => {
                const condition: any = {
                    search: {
                        name: {name: {test: "ttt"}}
                    }
                };
                const fn = () => sanitizeFilter(condition);

                expect(fn).not.toThrow();
                expect(fn()).toMatchSnapshot();
            });

            it("by Name condition", () => {
                const condition: any = {
                    search: {
                        name: {condition: {test: "ttt"}}
                    }
                };
                const fn = () => sanitizeFilter(condition);

                expect(fn).not.toThrow();
                expect(fn()).toMatchSnapshot();
            });
        });


        describe("Rating", () => {
            it("by Rating", () => {
                const condition: any = {
                    search: {
                        rating: "NaN"
                    }
                };
                const fn = () => sanitizeFilter(condition);

                expect(fn).not.toThrow();
                expect(fn()).toMatchSnapshot();
            });
            it("by Rating from", () => {
                const condition: any = {
                    search: {
                        rating: {from: "NaN", to: 5}
                    }
                };
                const fn = () => sanitizeFilter(condition);

                expect(fn).not.toThrow();
                expect(fn()).toMatchSnapshot();
            });
            it("by Rating to", () => {
                const condition: any = {
                    search: {
                        rating: {from: 4, to: "NaN"}
                    }
                };
                const fn = () => sanitizeFilter(condition);

                expect(fn).not.toThrow();
                expect(fn()).toMatchSnapshot();
            });
        });


        it("by Released", () => {
            const condition: any = {
                search: {
                    released: "BackIn2000"
                }
            };
            const fn = () => sanitizeFilter(condition);

            expect(fn).not.toThrow();
            expect(fn()).toMatchSnapshot();
        });
        it("by Status", () => {
            const condition: any = {
                search: {
                    status: "NotCancelled2"
                }
            };
            const fn = () => sanitizeFilter(condition);

            expect(fn).not.toThrow();
            expect(fn()).toMatchSnapshot();
        });
        it("by type", () => {
            const condition: any = {
                search: {
                    type: "InvalidType"
                }
            };
            const fn = () => sanitizeFilter(condition);

            expect(fn).not.toThrow();
            expect(fn()).toMatchSnapshot();
        });

    });
});
