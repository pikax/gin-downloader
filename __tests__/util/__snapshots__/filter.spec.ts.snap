// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter util invalid inputs Genre by Genre 1`] = `
Object {
  "search": Object {},
}
`;

exports[`filter util invalid inputs Genre by inGenre 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
  },
}
`;

exports[`filter util invalid inputs Genre by outGenre 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
  },
}
`;

exports[`filter util invalid inputs Name by Name 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
  },
}
`;

exports[`filter util invalid inputs Name by Name condition 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
  },
}
`;

exports[`filter util invalid inputs Rating by Rating 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
    "rating": Object {},
  },
}
`;

exports[`filter util invalid inputs Rating by Rating from 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
    "rating": Object {
      "to": 5,
    },
  },
}
`;

exports[`filter util invalid inputs Rating by Rating to 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
    "rating": Object {
      "from": 4,
    },
  },
}
`;

exports[`filter util invalid inputs by Artist 1`] = `
Object {
  "search": Object {
    "artist": Object {
      "condition": 1,
      "name": "",
    },
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
  },
}
`;

exports[`filter util invalid inputs by Author 1`] = `
Object {
  "search": Object {
    "author": Object {
      "condition": 1,
      "name": "",
    },
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
  },
}
`;

exports[`filter util invalid inputs by Mature 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
  },
}
`;

exports[`filter util invalid inputs by Released 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
    "released": Object {
      "condition": 0,
      "value": null,
    },
  },
}
`;

exports[`filter util invalid inputs by Status 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
  },
}
`;

exports[`filter util invalid inputs by type 1`] = `
Object {
  "search": Object {
    "genre": Object {
      "condition": 0,
      "inGenres": Array [],
      "outGenres": Array [],
    },
    "name": Object {
      "condition": 1,
      "name": "",
    },
  },
}
`;

exports[`filter util should sanitize filter: string 1`] = `
Object {
  "search": Object {
    "name": Object {
      "condition": 1,
      "name": "gintama",
    },
  },
}
`;

exports[`filter util should sanitize filter: user search  1`] = `
Object {
  "search": Object {
    "artist": Object {
      "condition": 1,
      "name": "Sorachi Hideaki",
    },
    "author": Object {
      "condition": 1,
      "name": "Sorachi Hideaki",
    },
    "genre": Object {
      "condition": 0,
      "inGenres": Array [
        "Shounen",
        "Action",
      ],
      "outGenres": Array [],
    },
    "mature": true,
    "name": Object {
      "condition": 1,
      "name": "gintama",
    },
    "rating": Object {
      "from": 1,
    },
    "released": Object {
      "condition": 0,
      "value": 2000,
    },
    "status": "Ongoing",
    "type": "Manga",
  },
}
`;
